### ISSUE TITLE
LAist: Push new LAssetHost plugin to production

### ISSUE THREAD

**Issue by arellaTV** [SCPR/SCPRv4#1229]
AC:

- [x] The new LAssethost is reviewed by Brian Frank
- [x] The new LAssethost is code reviewed
- [x] Githooks are set up to support pushing to production
- [x] The new LAssethost is shipped to production

**Comment by arellaTV** [comment]
Going to increase the estimate on this one as there are a few bugs caught as Brian Frank reviewed:
- [ ] Images don't show up at first when clicking AssetHost from the editor, but will show after logging in
- [ ] "Select Asset" and "View in Admin buttons" don't work
- [ ] "Save and Close" doesn't work

**Comment by arellaTV** [comment]
- [ ] Spin up LAist Assethost locally

**Comment by arellaTV** [comment]
Bug reported by Carla, Lisa, and Ryan:

- [ ] "not sure if this is an actual error  but i'm getting a failed to save asset error in laist's asset host, though it appears changes are in fact saving (edited)"

![](https://user-images.githubusercontent.com/16679701/42706083-751d3524-868a-11e8-9b39-95330cd6dd0a.png)

**Comment by arellaTV** [comment]
All bugfixes/features are scheduled for pushing tomorrow night.

**Comment by arellaTV** [comment]
Wrote a guide on how to use the new Asset Host plugin
https://docs.google.com/presentation/d/1vu-6EqdM4zHL1tz_JP6LS-FnfbyoyhA4CozztRSsVl0/edit#slide=id.p

**Comment by arellaTV** [comment]
Changing this ticket from a 2 to a 5 as I definitely underestimated

**Comment by arellaTV** [comment]
To do list for tonight's push
- [ ] create a snapshot
- [ ] perform upgrade
- [ ] grant permissions to correct folders
- [ ] push out mt plugins

**Comment by arellaTV** [comment]
Here's some diffs I collected
```bash
$ diff plugins/LAistMCE/ TinyMCE/
diff plugins/LAistMCE/config.yaml TinyMCE/config.yaml
1,2c1,2
< id: LAistMCE
< name: LAistMCE
---
> id: TinyMCE
> name: TinyMCE
4,5c4,5
< author_link: https://www.github.com/scpr/laist-mce
< author_name: LAist
---
> author_link: http://www.movabletype.org/
> author_name: Six Apart, Ltd.
7c7
< l10n_class: LAistMCE::L10N
---
> l10n_class: TinyMCE::L10N
11c11
<         label: LAistMCE
---
>         label: TinyMCE
Common subdirectories: plugins/LAistMCE/lib and TinyMCE/lib
Only in plugins/LAistMCE/: README.md
Common subdirectories: plugins/LAistMCE/tmpl and TinyMCE/tmpl
```

```bash
$ diff plugins/FormattedTextForLAistMCE/ FormattedTextForTinyMCE/
diff plugins/FormattedTextForLAistMCE/config.yaml FormattedTextForTinyMCE/config.yaml
1,7c1,7
< id: FormattedTextForLAistMCE
< name: FormattedTextForLAistMCE
< version: 1.4
< author_link: https://www.github.com/scpr/laist-mce
< author_name: LAist
< description: <MT_TRANS phrase='Add the "Insert Boilerplate" button to the LAistMCE'>
< l10n_class: FormattedTextForLAistMCE::L10N
---
> id: FormattedTextForTinyMCE
> name: FormattedTextForTinyMCE
> version: 1.3
> author_link: http://www.movabletype.org/
> author_name: Six Apart, Ltd.
> description: <MT_TRANS phrase='Add the "Insert Boilerplate" button to the TinyMCE.'>
> l10n_class: FormattedTextForTinyMCE::L10N
19c19
<                 code: $FormattedTextForLAistMCE::FormattedTextForLAistMCE::App::view_text
---
>                 code: $FormattedTextForTinyMCE::FormattedTextForTinyMCE::App::view_text
Common subdirectories: plugins/FormattedTextForLAistMCE/lib and FormattedTextForTinyMCE/lib
Common subdirectories: plugins/FormattedTextForLAistMCE/tmpl and FormattedTextForTinyMCE/tmpl
```

```bash
$ diff plugins/LAistMCE/ TinyMCE/
Common subdirectories: plugins/LAistMCE/lib and TinyMCE/lib
Common subdirectories: plugins/LAistMCE/tiny_mce and TinyMCE/tiny_mce
```

```bash
$ diff plugins/FormattedTextForLAistMCE/ FormattedTextForTinyMCE/
diff plugins/FormattedTextForLAistMCE/extension.js FormattedTextForTinyMCE/extension.js
3,25c3,12
<     var config  = MT.Editor.TinyMCE.config;
<     var wisiwyg_buttons = config.plugin_mt_wysiwyg_buttons1 + ',|,template';
<     var source_buttons  = (function(buttons) {
<         var replaced        = false;
<         var before          = /,asset_host/;
<         var template_button = ',|,mt_source_template';
< 
<         buttons = buttons.replace(before, function($0) {
<             replaced = true;
<             return $0 + template_button;
<         });
< 
<         if (! replaced) {
<             buttons += template_button;
<         }
< 
<         return buttons;
<     })(config.plugin_mt_source_buttons1);
< 
<     $.extend(config, {
<         plugins: config.plugins + ',template,mt_formatted_text',
<         plugin_mt_wysiwyg_buttons1: wisiwyg_buttons,
<         plugin_mt_source_buttons1: source_buttons
---
> var config  = MT.Editor.TinyMCE.config;
> var wisiwyg_buttons = config.plugin_mt_wysiwyg_buttons1 + ',|,template';
> var source_buttons  = (function(buttons) {
>     var replaced        = false;
>     var before          = /,mt_insert_image/;
>     var template_button = ',|,mt_source_template';
> 
>     buttons = buttons.replace(before, function($0) {
>         replaced = true;
>         return $0 + template_button;
28,35c15,35
<     tinymce.create('tinymce.plugins.MTFormattedText', {
<         init : function(ed, url) {
<             tinymce.addI18n(tinyMCE.settings.language, {
<                 'Insert template': trans('Insert Boilerplate'),
<                 'Templates': trans('Name'),
<             });
<         }
<     });
---
>     if (! replaced) {
>         buttons += template_button;
>     }
> 
>     return buttons;
> })(config.plugin_mt_source_buttons1);
> 
> $.extend(config, {
>     plugins: config.plugins + ',template,mt_formatted_text',
>     plugin_mt_wysiwyg_buttons1: wisiwyg_buttons,
>     plugin_mt_source_buttons1: source_buttons
> });
> 
> tinymce.create('tinymce.plugins.MTFormattedText', {
>     init : function(ed, url) {
>         tinymce.addI18n(tinyMCE.settings.language, {
>             'Insert template': trans('Insert Boilerplate'),
>             'Templates': trans('Name'),
>         });
>     }
> });
37c37
<     tinymce.PluginManager.add('mt_formatted_text', tinymce.plugins.MTFormattedText, ['template']);
---
> tinymce.PluginManager.add('mt_formatted_text', tinymce.plugins.MTFormattedText, ['template']);
39c39
<     })(jQuery);
---
> })(jQuery);
```

**Comment by arellaTV** [comment]
For the most part, it looks like the major difference between the new plugins and old ones are in `plugins/FormattedTextForLAistMCE/extension.js`, and that seems to be a client side script. The main issue behind the slowdowns is most likely server side as the average TTFB when loading an `Edit Entry` page is ~12s

![](https://user-images.githubusercontent.com/16679701/43265703-67a78584-909e-11e8-8ee7-2f054d29dc4d.png)

I checked the source of the `Edit Entry` and noticed that the entire collection of Tags are being loaded onto the page in order to make the autocomplete work. In the PERL script for edit entry, it starts by calling [get_tags_js](https://github.com/SCPR/movabletype/blob/f02afa4e1ed554ebc3c265d3856fb09e1666eb00/lib/MT/CMS/Entry.pm#L420-L421) which is defined [here](https://github.com/SCPR/movabletype/blob/f02afa4e1ed554ebc3c265d3856fb09e1666eb00/lib/MT/Tag.pm#L605-L626).

I experimented in staging and disabled the `get_tags_js` call, and it reduced load time by 70%. Autocomplete stops working, but adding tags still works. I think the same optimization done on production could yield similar results.
